_drafts/2024-07-25-testing-raise.md

```diff
@@ -0,0 +1,531 @@
+                createPortfolio(CreatePortfolio(UserId("bob"), Money(1000.0)))
+            }
+        }
+    actualResult.shouldBeLeft(GenericError(exception))
+}
+```
+
+The above test verifies the behavior of the use case when there was an unexpected error during the execution of the port. We used the `answers` function, the most generic one available in MockK.
+
+For completeness, we can translate the above test using Mockito to understand the differences between the two libraries. In detail, we'll use the library `mockito-kotlin` on top of Mockito to have a more idiomatic look and feel:
+
+```kotlin
+@Test
+fun `given a userId and an initial amount, when executed with error, then propagates the error properly`() {
+    val exception = RuntimeException("Ooops!")
+    val actualResult =
+        either {
+            val countUserPortfoliosPort =
+                mock<CountUserPortfoliosPort> {
+                    on { 
+                        countByUserId(UserId("bob")) 
+                    } doAnswer { 
+                        raise(GenericError(exception)) 
+                    }
+                }
+            with(createPortfolioUseCase(countUserPortfoliosPort)) {
+                createPortfolio(CreatePortfolio(UserId("bob"), Money(1000.0)))
+            }
+        }
+    EitherAssert.assertThat(actualResult).containsOnLeft(GenericError(exception))
+}
+```
+
+As we can see, the only part that has changed is the DSL offered by the mocking library. The rest of the test remains more or less the same.
+
+## 4. Conclusions
+
+In this article, we saw how to test a function declared in a `Raise<E>` context. We introduced several approaches and used many libraries. We saw how to write a test using JUnit 5 and the Kotest testing frameworks. We saw how to use the `assertj-arrow-core` library to test functions using the Raise DSL. Then, we translated the same tests using the Kotest extension asserts for Arrow. Then, we focused on mocking. The first solution we approached was using fake objects. Despite some initial programming, the fake objects are natural and ergonomic. We also saw how to mock a function declared in a `Raise<E>` context using the MockK and Mockito libraries. Mocking feels less idiomatic and more complex due to the `Raise<E>` context, which neither libraries manage natively. As of today, we still miss some Mockito's extensions functions to handle the Raise DSL natively, which would make the mocking process more straightforward.
+
+## 5. Appendix
+
+As we said, Kotlin context receivers were deprecated in more recent versions of Kotlin. For the above reason, we preferred to use something other than context receivers in this article. However, it's not debatable whether the syntax of context receivers is clean and concise. We introduced it in the article [Kotlin Context Receivers: A Comprehensive Guide](https://blog.rockthejvm.com/kotlin-context-receivers/). Let's see how our example changes if we use context receivers.
```

Comment:
"it's not debatable whether the syntax is clean and concise" 
What do you mean? Is it not up for debate... meaning it's very good, or it's very bad? 😅

If you want to say that context receivers "objectively" make the syntax better, you can say something like "context receivers almost invariably make the code cleaner and more concise".